### PR TITLE
Enzo: building extra tools, optimization flag and gcc 10.2 fix

### PR DESCRIPTION
**Pull request by nsimakov**
This PR has three things

1. Built extra tools, namely inits and ring. Both needed for simulations.
2. Added optimization flag many new compilers do not produce stable code. I.e. with high and aggressive optimization some things diverges.
  *. In my test case gcc 9.3 and 10.2 failed to complete simulation if compiled with high or aggressive optimization
3. A fix for compiling with gcc 10.2 fix, original error message:
```
Error: Type mismatch in argument 'ispec' at (1); passed INTEGER(4) to INTEGER(8)
```

-Nikolay

PS: same as https://github.com/spack/spack/pull/25397 , was having some difficulties with PR system

### CODE REVIEW

File: var/spack/repos/builtin/packages/enzo/package.py [hunk 1]
```diff
@@ -53,16 +62,29 @@ def edit(self, spec, prefix):
             filter_file('^LOCAL_HYPRE_INSTALL.*',
                         'LOCAL_HYPRE_INSTALL =',
                         'Make.mach.spack')
+            if spec.satisfies('%gcc@10.2.0:'):
+                filter_file(r'^MACH_FFLAGS\s*=\s*',
+                            'MACH_FFLAGS   = -fallow-argument-mismatch ',
+                            'Make.mach.spack')
+                filter_file(r'^MACH_F90FLAGS\s*=\s*',
+                            'MACH_F90FLAGS = -fallow-argument-mismatch ',
+                            'Make.mach.spack')
```
Comment (michaelkuhn) [inline var/spack/repos/builtin/packages/enzo/package.py]: Why is this needed? We already set the flag in `flag_handler` above.

File: var/spack/repos/builtin/packages/enzo/package.py [hunk 2]
```diff
@@ -53,16 +62,29 @@ def edit(self, spec, prefix):
...
+        with working_dir('src/ring'):
+            make()
 
     def install(self, spec, prefix):
         install_tree('bin', prefix.bin)
         install_tree('doc', prefix.doc)
         install_tree('input', prefix.input)
         install_tree('run', prefix.run)
+        with working_dir(join_path('src', 'ring')):
+            install('ring.exe', join_path(prefix.bin, 'ring'))
```
Comment (michaelkuhn) [inline var/spack/repos/builtin/packages/enzo/package.py]: Why not simply do the following?
```python
install('src/ring/ring.exe', join_path(prefix.bin, 'ring'))
```

**Review by michaelkuhn — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/enzo/package.py (hunk 1 above)
Comment (nsimakov) [reply var/spack/repos/builtin/packages/enzo/package.py]: Well, I really don't know why, but in my setup with gcc 10.2 and without explicit modifications to MACH_F90FLAGS or MACH_FFLAGS it does not work. Got "`Error: Type mismatch in argument 'ispec' at (1); passed INTEGER(4) to INTEGER(8)`"

May be `flag_handler` is not good enough for custom enzo makefile code which seems to use several dialects of fortran at a same time.

File: var/spack/repos/builtin/packages/enzo/package.py (hunk 2 above)
Comment (nsimakov) [reply var/spack/repos/builtin/packages/enzo/package.py]: Huh, I think it should work. But I don't have time to test it right now, can we skip on it?

File: var/spack/repos/builtin/packages/enzo/package.py (hunk 1 above)
Comment (michaelkuhn) [reply var/spack/repos/builtin/packages/enzo/package.py]: That's weird. I just removed this part and it still compiles for me (with GCC 11). I think it would be good to find out what's going on there.

File: var/spack/repos/builtin/packages/enzo/package.py (hunk 2 above)
Comment (michaelkuhn) [reply var/spack/repos/builtin/packages/enzo/package.py]: I also built Enzo with this change and it works.

File: var/spack/repos/builtin/packages/enzo/package.py (hunk 1 above)
Comment (nsimakov) [reply var/spack/repos/builtin/packages/enzo/package.py]: Weird, flag_handler is working for me too now. I did rebase prior PR may be something was change meanwhile. Anyway I removed it. After rebasing and running another installation i will update PR.

**Comment by nsimakov** [comment]
Thanks, for help

**Review by michaelkuhn — APPROVED** [review]

**Comment by michaelkuhn** [comment]
@spackbot run pipeline

**Comment by michaelkuhn** [comment]
@spackbot run pipeline
